File: app/src/main/java/org/wikipedia/readinglist/RemoveFromReadingListsDialog.java
package org.wikipedia.readinglist;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;

import org.wikipedia.R;
import org.wikipedia.readinglist.database.ReadingList;
import org.wikipedia.readinglist.database.ReadingListDbHelper;
import org.wikipedia.readinglist.database.ReadingListPage;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.apache.commons.collections4.CollectionUtils.isEmpty;
import static org.apache.commons.collections4.IterableUtils.first;
import static org.apache.commons.collections4.IterableUtils.get;
import static org.apache.commons.collections4.IterableUtils.size;
import static org.apache.commons.collections4.ListUtils.emptyIfNull;

public class RemoveFromReadingListsDialog {
    public interface Callback {
        void onDeleted(@NonNull List<ReadingList> lists, @NonNull ReadingListPage page);
    }

    @Nullable private List<ReadingList> listsContainingPage;

    public RemoveFromReadingListsDialog(@NonNull List<ReadingList> listsContainingPage) {
        this.listsContainingPage = listsContainingPage;
        ReadingList.sort(listsContainingPage, ReadingList.SORT_BY_NAME_ASC);
    }

    public void deleteOrShowDialog(@NonNull Context context, @Nullable Callback callback) {
        if (isEmpty(listsContainingPage)) {
            return;
        }
        if (size(listsContainingPage) == 1) {
            final ReadingList first = first(listsContainingPage);
            if (!first.pages().isEmpty()) {
                ReadingListDbHelper.instance().markPagesForDeletion(first,
                        Collections.singletonList(first.pages().get(0)));
                if (callback != null) {
                    callback.onDeleted(emptyIfNull(listsContainingPage), first.pages().get(0));
                }
            }
            return;
        }
        showDialog(context, callback);
    }

    private void showDialog(@NonNull Context context, @Nullable final Callback callback) {
        final String[] listNames = new String[size(listsContainingPage)];
        final boolean[] selected = new boolean[listNames.length];

        for (int i = 0; i < size(listsContainingPage); i++) {
            listNames[i] = get(listsContainingPage, i).title();
        }

        new AlertDialog.Builder(context)
                .setTitle(R.string.reading_list_remove_from_lists)
                .setPositiveButton(R.string.reading_list_remove_list_dialog_ok_button_text, (dialog, which) -> {
                    boolean atLeastOneSelected = false;
                    List<ReadingList> newLists = new ArrayList<>();
                    for (int i = 0; i < listNames.length; i++) {
                        final ReadingList list = get(listsContainingPage, i);
                        if (selected[i]) {
                            atLeastOneSelected = true;
                            ReadingListDbHelper.instance().markPagesForDeletion(list,
                                    Collections.singletonList(list.pages().get(0)));
                            newLists.add(list);
                        }
                    }
                    if (callback != null && atLeastOneSelected) {
                        callback.onDeleted(newLists, first(listsContainingPage).pages().get(0));
                    }
                })
                .setNegativeButton(R.string.reading_list_remove_from_list_dialog_cancel_button_text, null)
                .setMultiChoiceItems(listNames, selected, (dialog, which, checked) -> selected[which] = checked)
                .create()
                .show();
    }
}

File: app/src/main/java/org/wikipedia/readinglist/RemoveFromReadingListsDialog.kt
package org.wikipedia.readinglist

import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.launch
import org.wikipedia.R
import org.wikipedia.database.AppDatabase
import org.wikipedia.readinglist.database.ReadingList
import org.wikipedia.readinglist.database.ReadingListPage
import org.wikipedia.util.log.L

class RemoveFromReadingListsDialog(private val listsContainingPage: List<ReadingList>) {
    fun interface Callback {
        fun onDeleted(lists: List<ReadingList>, page: ReadingListPage)
    }

    init {
        ReadingList.sort(listsContainingPage as MutableList<ReadingList>, ReadingList.SORT_BY_NAME_ASC)
    }

    fun deleteOrShowDialog(activity: AppCompatActivity, callback: Callback?) {
        if (listsContainingPage.isEmpty()) {
            return
        }
        if (listsContainingPage.size == 1 && listsContainingPage[0].pages.isNotEmpty()) {
            activity.lifecycleScope.launch(CoroutineExceptionHandler { _, throwable ->
                L.w(throwable)
            }) {
                AppDatabase.instance.readingListPageDao().markPagesForDeletion(listsContainingPage[0], listOf(listsContainingPage[0].pages[0]))
                callback?.onDeleted(listsContainingPage, listsContainingPage[0].pages[0])
            }
            return
        }
        showDialog(activity, callback)
    }

    private fun showDialog(activity: AppCompatActivity, callback: Callback?) {
        val selectedLists = BooleanArray(listsContainingPage.size)
        MaterialAlertDialogBuilder(activity)
            .setTitle(R.string.reading_list_remove_from_lists)
            .setPositiveButton(R.string.reading_list_remove_list_dialog_ok_button_text) { _, _ ->
                activity.lifecycleScope.launch(CoroutineExceptionHandler { _, throwable ->
                    L.w(throwable)
                }) {
                    val newLists = (listsContainingPage zip selectedLists.asIterable())
                        .filter { (_, selected) -> selected }
                        .map { (listContainingPage, _) ->
                            AppDatabase.instance.readingListPageDao().markPagesForDeletion(
                                listContainingPage,
                                listOf(listContainingPage.pages[0])
                            )
                            listContainingPage
                        }
                    if (newLists.isNotEmpty()) {
                        callback?.onDeleted(newLists, listsContainingPage[0].pages[0])
                    }
                }
            }
            .setNegativeButton(R.string.reading_list_remove_from_list_dialog_cancel_button_text, null)
            .setMultiChoiceItems(listsContainingPage.map { it.title }.toTypedArray(), selectedLists) { _, which, checked ->
                selectedLists[which] = checked
            }
            .show()
    }
}
